net/pfSense-pkg-ntopng/files/usr/local/pkg/ntopng.inc

```diff
@@ -217,112 +325,208 @@ function ntopng_redis_started() {
-
-	if (ntopng_redis_started()) {
-		/* Delete all the keys of all the existing Redis databases */
-		mwexec("{$redis_path}/redis-cli flushall");
-		log_error(gettext("[ntopng] Flushed Redis DB."));
-		/* Set admin password while redis-server is still running */
-		ntopng_set_redis_password();
-		log_error(gettext("[ntopng] Set admin password for Redis DB."));
-		/* Stop services and delete all graphs, data and dump flows */
-		ntopng_services_stop();
-		if (is_dir("/var/db/ntopng/")) {
-			mwexec("rm -rf /var/db/ntopng/");
-			log_error(gettext("[ntopng] Deleted ntopng traffic data and graphs."));
+function ntopng_delete_data() {
+	global $data_dir, $redis_cli;
+
+	log_error(gettext("[ntopng] Request to delete data."));
+
+	/* Shut down ntopng */
+	ntopng_stop_ntopng();
+
+	/* Ensure redis is running */
+	if (ntopng_start_redis()) {
+		/* Delete redis cache entries */
+		mwexec("{$redis_cli} --scan --pattern '*.cache.*' | xargs -L 100 {$redis_cli} DEL");
+		log_error(gettext("[ntopng] Deleted cache entries in Redis DB."));
+
+		/* Stop redis */
+		ntopng_stop_redis();
+
+		/* delete all rrd and db files and remove empty directories */
+		if (is_dir($data_dir)) {
+			mwexec("/usr/bin/find {$data_dir} -name '*.db' -o -name '*.rrd' | /usr/bin/xargs /bin/rm");
+			mwexec("/usr/bin/find {$data_dir} -depth -type d | /usr/bin/xargs /bin/rmdir", true);
+			log_error(gettext("[ntopng] Deleted ntopng traffic data and alerts."));
 		} else {
-			log_error(gettext("[ntopng] Nothing to delete; /var/db/ntopng/ directory not found."));
+			log_error(gettext("[ntopng] Nothing to delete; directory {$data_dir} not found."));
 		}
-		/* Re-create the required directory structure with proper permissions */
-		ntopng_create_datadir();
-		log_error(gettext("[ntopng] Re-created required data directory structure."));
-		/* Resync settings and restart services if enabled */
+
+		/* Resync settings and restart service if enabled */
 		unset($_POST['Delete']);
 		ntopng_sync_package();
 		log_error(gettext("[ntopng] Resynced ntopng settings."));
 	} else {
-		$error = "Cannot delete data - redis-server is not running.";
-		log_error(gettext("[ntopng] {$error}"));
-		file_notice("ntopng", $error, "ntopng Delete Data", "");
+		log_error(gettext("[ntopng] Cannot delete data - redis-server is not running."));
 	}
 }
 
+function ntopng_reset() {
+	global $data_dir, $redis_cli;
+
+	log_error(gettext("[ntopng] Request to perform a full reset."));
+
+	/* Shut everything down */
+	ntopng_stop_service();
+
+	/* Remove the entire data directory */
+	mwexec("/bin/rm -rf {$data_dir}");
+	log_error(gettext("[ntopng] All traffic data, alerts, preferences and customization deleted."));
+
+	/* Resync settings and restart service if enabled */
+	unset($_POST['Delete']);
+	ntopng_sync_package();
+	log_error(gettext("[ntopng] Resynced ntopng settings."));
+}
+
 function ntopng_validate_input($post, &$input_errors) {
+	/* Validate password */
```

Comment:
As long as you're validating the password, add in something like this as well:

```php
$input_errors = [];
//...
if (function_exists('validate_password')) {
	$input_errors = array_merge($input_errors, validate_password('redis', $post['redis_password']));
}
```

That will reject password values considered invalid by the base system.

Reply:
I don't see validate_password() used in any packages, nor was I able to find it in pfSense itself. Am I missing something?

FWIW, I am very reluctant to introduce potentially new password requirements. This could result in a previously set admin password suddenly being unacceptable, which would prevent other settings from being changed. I think it's better to leave this alone. If ntopng had its own externalized validator I might feel differently, but it doesn't.

Reply:
It would only be preventing values nobody should be using in the first place.

https://redmine.pfsense.org/issues/15266

https://github.com/pfsense/pfsense/blob/master/src/etc/inc/auth.inc#L652

At the moment on CE it would just stop someone from using `redis` for the password, and on Plus it would also prevent using `pfsense`.

It makes it consistent with the base system password behavior on Plus 24.03 and later / CE 2.8.0 and later. The function doesn't exist on Plus <= 23.09.1 / CE <=2.7.2 so it's a no-op there.

Reply:
Okay.

What is the signature of validate_password()?